odl/space/fspace.py

```diff
@@ -6,371 +6,1277 @@
 from future import standard_library
 standard_library.install_aliases()
 from builtins import super
 
-from inspect import isfunction
+import inspect
 import numpy as np
+import sys
+import warnings
 
-from odl.operator.operator import Operator, _dispatch_call_args
-from odl.set import (RealNumbers, ComplexNumbers, Set, Field, LinearSpace,
-                     LinearSpaceElement)
+from odl.set import RealNumbers, ComplexNumbers, Set, LinearSpace
+from odl.set.space import LinearSpaceElement
 from odl.util import (
-    is_real_dtype, is_complex_floating_dtype, dtype_repr,
-    complex_dtype, real_dtype,
+    is_real_dtype, is_complex_floating_dtype, dtype_repr, dtype_str,
+    complex_dtype, real_dtype, signature_string,
     is_valid_input_array, is_valid_input_meshgrid,
-    out_shape_from_array, out_shape_from_meshgrid, vectorize)
-from odl.util.utility import preload_first_arg
+    out_shape_from_array, out_shape_from_meshgrid, vectorize, broadcast_to,
+    writable_array)
+from odl.util.utility import preload_first_arg, getargspec
 
 
-__all__ = ('FunctionSet', 'FunctionSetElement',
-           'FunctionSpace', 'FunctionSpaceElement')
+__all__ = ('FunctionSpace',)
+
+
+def _check_out_arg(func):
+    """Check if ``func`` has an (optional) ``out`` argument.
+
+    Also verify that the signature of ``func`` has no ``*args`` since
+    they make argument propagation a hassle.
+
+    Parameters
+    ----------
+    func : callable
+        Object that should be inspected.
+
+    Returns
+    -------
+    has_out : bool
+        ``True`` if the signature has an ``out`` argument, ``False``
+        otherwise.
+    out_is_optional : bool
+        ``True`` if ``out`` is present and optional in the signature,
+        ``False`` otherwise.
+
+    Raises
+    ------
+    TypeError
+        If ``func``'s signature has ``*args``.
+    """
+    if sys.version_info.major > 2:
+        spec = inspect.getfullargspec(func)
+        kw_only = spec.kwonlyargs
+    else:
+        spec = inspect.getargspec(func)
+        kw_only = ()
+
+    if spec.varargs is not None:
+        raise TypeError('*args not allowed in function signature')
+
+    pos_args = spec.args
+    pos_defaults = () if spec.defaults is None else spec.defaults
+
+    has_out = 'out' in pos_args or 'out' in kw_only
+    if 'out' in pos_args:
+        has_out = True
+        out_is_optional = (
+            pos_args.index('out') >= len(pos_args) - len(pos_defaults))
+    elif 'out' in kw_only:
+        has_out = out_is_optional = True
+    else:
+        has_out = out_is_optional = False
+
+    return has_out, out_is_optional
 
 
 def _default_in_place(func, x, out, **kwargs):
     """Default in-place evaluation method."""
-    out[:] = func(x, **kwargs)
+    result = np.array(func._call_out_of_place(x, **kwargs), copy=False)
+    if result.dtype == object:
+        # Different shapes encountered, need to broadcast
+        flat_results = result.ravel()
```

Comment:
I'll check, but I think the order is fixed here, namely determined by the input `x`, which is usually a meshgrid. Hence `'C'` order is correct, since it corresponds to the coordinate axes in `np.broadcast(*x)`.

Reply:
Ah no, this is entirely internal. We flatten the object array holding the results (broadcasting needed), then we broadcast the results to final shape and create a new array from those arrays. That array is still flat in the output dimensions, so we reshape to `out_shape + scalar_shape`.
As long as that last reshape uses the same order as the initial ravel, things are consistent.

Reply:
I'll add some comments, but honestly this part was a lot of trial-and-error since there is just so much stuff going on that it's hard to keep track of all the possible reshaping and broadcasting cases. I trust the unit tests here, they are fairly comprehensive.

Reply:
Agree that "proof by unit test", while perhaps not satisfying, is quite OK. Add a comment and we're fine